Brokerages/Zerodha/ZerodhaSubscriptionManager.cs

```diff
@@ -0,0 +1,640 @@
+        {
+            Log.Error($"ZerodhaSubscriptionManager.OnError(): Message: {e.Message} Exception: {e.Exception}");
+        }
+
+
+        private void OnMessage(object sender, MessageData e)
+        {
+           
+            _timerTick = _interval;
+            if (e.MessageType == WebSocketMessageType.Binary)
+            {
+                if (e.Count == 1)
+                {
+                    if (_debug) Console.WriteLine(DateTime.Now.ToLocalTime() + " Heartbeat");
+                }
+                else
+                {
+                    int offset = 0;
+                    ushort count = ReadShort(e.Data, ref offset); //number of packets
+                    if (_debug) Console.WriteLine("No of packets: " + count);
+                    if (_debug) Console.WriteLine("No of bytes: " + e.Count);
+
+                    for (ushort i = 0; i < count; i++)
+                    {
+                        ushort length = ReadShort(e.Data, ref offset); // length of the packet
+                        if (_debug) Console.WriteLine("Packet Length " + length);
+                        Messages.Tick tick = new Messages.Tick();
+                        if (length == 8) // ltp
+                            tick = ReadLTP(e.Data, ref offset);
+                        else if (length == 28) // index quote
+                            tick = ReadIndexQuote(e.Data, ref offset);
+                        else if (length == 32) // index quote
+                            tick = ReadIndexFull(e.Data, ref offset);
+                        else if (length == 44) // quote
+                            tick = ReadQuote(e.Data, ref offset);
+                        else if (length == 184) // full with marketdepth and timestamp
+                            tick = ReadFull(e.Data, ref offset);
+                        // If the number of bytes got from stream is less that that is required
+                        // data is invalid. This will skip that wrong tick
+                        if (tick.InstrumentToken != 0 && offset <= e.Count)
+                        {
+                            var sym = _subscriptionsById[tick.InstrumentToken];
+                            //TODO handle this
+                            //EmitQuoteTick(sym, upd.bPr, tick.Bids.bSz, upd.aPr, upd.aSz);
+                        }
+                    }
+                }
+            }
+            else if (e.MessageType == WebSocketMessageType.Text)
+            {
+                string message = Encoding.UTF8.GetString(e.Data.Take(e.Count).ToArray());
+                if (_debug) Console.WriteLine("WebSocket Message: " + message);
+
+                Dictionary<string, dynamic> messageDict = Utils.JsonDeserialize(message);
+                if (messageDict["type"] == "order")
+                {
+                    //TODO handle this
+                    //OnOrderUpdate?.Invoke(new Order(messageDict["data"]));
+                    //EmitFillOrder();
```

Comment:
Handling of order fills should be done using a dedicated websocket connection in the brokerage class so it is independent of the data feed subscriptions (e.g. when using the QC cloud environment this class wouldn't even be instantiated)

Reply:
Got it!